### ISSUE TITLE
drachtio server sends ACK to wrong destination port

### ISSUE THREAD

**Issue by spointer** [davehorton/drachtio-server#100]
Hello Dave,

I have configured drachtio-app which proxy to another drachtio-app(UAS)

srf.invite((req, res) => {
    srf.proxyRequest(req, '121.160.50.106:3520', {
        forking: 'sequential',
        remainInDialog: true,
        provisionalTimeout: '10s',
        finalTimeout: '30s',
        followRedirects: false
    }).then((results) => {
    });
});

All SIP methods are proxied well, but ACK is sent to the wrong destination port and as a result, ACK timeout is raised.
Proxy drachtio-server logs are as follows.
INVITE is proxied to the right destination(ie, the destination of the proxyRequest function) but ACK is proxied to the wrong port.

2019-10-15 01:21:37.368010 recv 888 bytes from udp/[111.111.11.247]:5060 at 01:21:37.367653:
INVITE sip:07080942391@121.160.50.106:5060;user=phone SIP/2.0
Via: SIP/2.0/UDP 111.111.11.247:5060;branch=z9hG4bK1d3e92ab45da51f21-cf919-0_850201
Via: SIP/2.0/UDP 111.111.11.180:5060;branch=z9hG4bK1d3e92ab45da51f21-cf919-0
From: <sip:01000000000@111.111.11.247>;tag=6775e234cb9ac685411b19510d3f9b00
To: <sip:07080942391@121.160.50.106>
Record-Route: <sip:111.111.11.247;lr>
Call-ID: 111.111.11.247_AAQvpn__DyEAABey0-lBRw--c14076749@a.com
CSeq: 850201 INVITE
Contact: <sip:01000000000@111.111.11.247:5060;user=phone>
Max-Forwards: 69
Content-Type: application/sdp
Content-Length: 287

v=0
o=- 1771508041 1771508041 IN IP4 111.233.26.249
s=ENSResip
c=IN IP4 111.233.26.249
t=0 0
m=audio 20636 RTP/AVP 0 8 4 18 101
a=fmtp:101 0-11
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:4 G723/8000
a=rtpmap:18 G729/8000
a=rtpmap:101 telephone-event/8000
a=sendrecv

2019-10-15 01:21:37.368573 send 459 bytes to udp/[111.111.11.247]:5060 at 01:21:37.368329:
SIP/2.0 100 Trying
Via: SIP/2.0/UDP 111.111.11.247:5060;branch=z9hG4bK1d3e92ab45da51f21-cf919-0_850201;rport=5060
Via: SIP/2.0/UDP 111.111.11.180:5060;branch=z9hG4bK1d3e92ab45da51f21-cf919-0
Record-Route: <sip:111.111.11.247;lr>
From: <sip:01000000000@111.111.11.247>;tag=6775e234cb9ac685411b19510d3f9b00
To: <sip:07080942391@121.160.50.106>
Call-ID: 111.111.11.247_AAQvpn__DyEAABey0-lBRw--c14076749@a.com
CSeq: 850201 INVITE
Content-Length: 0

2019-10-15 01:21:37.375207 ClientTransaction::setState - NOT STARTED --> CALLING 111.111.11.247_AAQvpn__DyEAABey0-lBRw--c14076749@a.com
2019-10-15 01:21:37.380255 send 1038 bytes to udp/[121.160.50.106]:3520 at 01:21:37.380008:
INVITE sip:07080942391@121.160.50.106:5060;user=phone SIP/2.0
Via: SIP/2.0/UDP 121.160.50.106:3417;rport;branch=z9hG4bK0abeccef-643e-47fa-a3a4-c528cc55de83
Via: SIP/2.0/UDP 111.111.11.247:5060;branch=z9hG4bK1d3e92ab45da51f21-cf919-0_850201;rport=5060
Via: SIP/2.0/UDP 111.111.11.180:5060;branch=z9hG4bK1d3e92ab45da51f21-cf919-0
Record-Route: <sip:121.160.50.106:3417;lr>
Record-Route: <sip:111.111.11.247;lr>
Max-Forwards: 68
From: <sip:01000000000@111.111.11.247>;tag=6775e234cb9ac685411b19510d3f9b00
To: <sip:07080942391@121.160.50.106>
Call-ID: 111.111.11.247_AAQvpn__DyEAABey0-lBRw--c14076749@a.com
CSeq: 850201 INVITE
Contact: <sip:01000000000@111.111.11.247:5060;user=phone>
Content-Type: application/sdp
Content-Length: 287

v=0
o=- 1771508041 1771508041 IN IP4 111.233.26.249
s=ENSResip
c=IN IP4 111.233.26.249
t=0 0
m=audio 20636 RTP/AVP 0 8 4 18 101
a=fmtp:101 0-11
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:4 G723/8000
a=rtpmap:18 G729/8000
a=rtpmap:101 telephone-event/8000
a=sendrecv

2019-10-15 01:21:37.387326 recv 624 bytes from udp/[121.160.50.106]:3520 at 01:21:37.387165:
SIP/2.0 100 Trying
Via: SIP/2.0/UDP 121.160.50.106:3417;rport=35398;branch=z9hG4bK0abeccef-643e-47fa-a3a4-c528cc55de83;received=172.21.0.1
Via: SIP/2.0/UDP 111.111.11.247:5060;branch=z9hG4bK1d3e92ab45da51f21-cf919-0_850201;rport=5060
Via: SIP/2.0/UDP 111.111.11.180:5060;branch=z9hG4bK1d3e92ab45da51f21-cf919-0
Record-Route: <sip:121.160.50.106:3417;lr>
Record-Route: <sip:111.111.11.247;lr>
From: <sip:01000000000@111.111.11.247>;tag=6775e234cb9ac685411b19510d3f9b00
To: <sip:07080942391@121.160.50.106>
Call-ID: 111.111.11.247_AAQvpn__DyEAABey0-lBRw--c14076749@a.com
CSeq: 850201 INVITE
Content-Length: 0

2019-10-15 01:21:37.387504 ClientTransaction::setState - CALLING --> PROCEEDING 111.111.11.247_AAQvpn__DyEAABey0-lBRw--c14076749@a.com
2019-10-15 01:21:38.632745 recv 970 bytes from udp/[121.160.50.106]:3520 at 01:21:38.632427:
SIP/2.0 200 OK
Via: SIP/2.0/UDP 121.160.50.106:3417;rport=35398;branch=z9hG4bK0abeccef-643e-47fa-a3a4-c528cc55de83;received=172.21.0.1
Via: SIP/2.0/UDP 111.111.11.247:5060;branch=z9hG4bK1d3e92ab45da51f21-cf919-0_850201;rport=5060
Via: SIP/2.0/UDP 111.111.11.180:5060;branch=z9hG4bK1d3e92ab45da51f21-cf919-0
Record-Route: <sip:121.160.50.106:3417;lr>
Record-Route: <sip:111.111.11.247;lr>
From: <sip:01000000000@111.111.11.247>;tag=6775e234cb9ac685411b19510d3f9b00
To: <sip:07080942391@121.160.50.106>;tag=2K0jKKeav0eSS
Call-ID: 111.111.11.247_AAQvpn__DyEAABey0-lBRw--c14076749@a.com
CSeq: 850201 INVITE
Contact: <sip:172.21.0.2:3520>
Content-Type: application/sdp
Content-Length: 267

v=0
o=- 1771508041 1771508043 IN IP4 121.160.50.106
s=coginsight
c=IN IP4 121.160.50.106
t=0 0
m=audio 10032 RTP/AVP 0 8 101
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-16
a=ptime:20
a=maxptime:150
a=sendrecv

2019-10-15 01:21:38.633436 ClientTransaction::setState - PROCEEDING --> TERMINATED 111.111.11.247_AAQvpn__DyEAABey0-lBRw--c14076749@a.com
2019-10-15 01:21:38.634079 send 849 bytes to udp/[111.111.11.247]:5060 at 01:21:38.633859:
SIP/2.0 200 OK
Via: SIP/2.0/UDP 111.111.11.247:5060;branch=z9hG4bK1d3e92ab45da51f21-cf919-0_850201;rport=5060
Via: SIP/2.0/UDP 111.111.11.180:5060;branch=z9hG4bK1d3e92ab45da51f21-cf919-0
Record-Route: <sip:121.160.50.106:3417;lr>
Record-Route: <sip:111.111.11.247;lr>
From: <sip:01000000000@111.111.11.247>;tag=6775e234cb9ac685411b19510d3f9b00
To: <sip:07080942391@121.160.50.106>;tag=2K0jKKeav0eSS
Call-ID: 111.111.11.247_AAQvpn__DyEAABey0-lBRw--c14076749@a.com
CSeq: 850201 INVITE
Contact: <sip:172.21.0.2:3520>
Content-Type: application/sdp
Content-Length: 267

v=0
o=- 1771508041 1771508043 IN IP4 121.160.50.106
s=coginsight
c=IN IP4 121.160.50.106
t=0 0
m=audio 10032 RTP/AVP 0 8 101
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-16
a=ptime:20
a=maxptime:150
a=sendrecv

2019-10-15 01:21:38.721960 recv 514 bytes from udp/[111.111.11.247]:5060 at 01:21:38.721626:
ACK sip:07080942391@121.160.50.106:5060;user=phone SIP/2.0
Via: SIP/2.0/UDP 111.111.11.247:5060;branch=z9hG4bK1d3e92ab45da51f21-cf919-1_850201
Via: SIP/2.0/UDP 111.111.11.180:5060;branch=z9hG4bK1d3e92ab45da51f21-cf919-1
From: <sip:01000000000@111.111.11.247>;tag=6775e234cb9ac685411b19510d3f9b00
To: <sip:07080942391@121.160.50.106>;tag=2K0jKKeav0eSS
Call-ID: 111.111.11.247_AAQvpn__DyEAABey0-lBRw--c14076749@a.com
CSeq: 850201 ACK
Route: <sip:121.160.50.106;lr>
Max-Forwards: 69
Content-Length: 0

**2019-10-15 01:21:38.725911 send 579 bytes to udp/[121.160.50.106]:5060 at 01:21:38.725738:**
ACK sip:07080942391@121.160.50.106:5060;user=phone SIP/2.0
Via: SIP/2.0/UDP 121.160.50.106:3417;rport;branch=z9hG4bK.gBajD8mNgmQ1p22j3g0Qy78SpH
Via: SIP/2.0/UDP 111.111.11.247:5060;branch=z9hG4bK1d3e92ab45da51f21-cf919-1_850201;rport=5060
Via: SIP/2.0/UDP 111.111.11.180:5060;branch=z9hG4bK1d3e92ab45da51f21-cf919-1
From: <sip:01000000000@111.111.11.247>;tag=6775e234cb9ac685411b19510d3f9b00
To: <sip:07080942391@121.160.50.106>;tag=2K0jKKeav0eSS
Call-ID: 111.111.11.247_AAQvpn__DyEAABey0-lBRw--c14076749@a.com
CSeq: 850201 ACK
Max-Forwards: 69
Content-Length: 0

**Comment by davehorton** [comment]
yes, it looks like the proxy is not behaving that well in a nat scenario.  I made some fixes for this for B2B functionality, but I may have to look into the same for proxy.

To summarize, here is what happens:

- server receives INVITE and proxies it on to udp/[121.160.50.106]:3520 
- server receives 200 OK with Contact header sip:172.21.0.2:3520, but the message comes from the IP 121.160.50.106 so we can see it is a nat;'ed server
- server receives an ACK, and sends it on to the same IP that it sent the INVITE, which is wrong.  It should either been sent to the Contact header in the 200 OK (172.21.0.2:3520), or if it detected that server was behind a nat, to 121.160.50.106:3520.

Which did you expect the ACK to be sent to?

**Comment by spointer** [comment]
exactly, you're right and i expected the ACK should be sent to 121.160.50.106:4520

**Comment by spointer** [comment]
Hi Dave,
Is there any progress of this issue?
I can provide the log turned on sofia log level to 9.

**Comment by davehorton** [comment]
Thanks for reminding me; meant to get to this last week but forgot.  Should have something to test this week.

**Comment by davehorton** [comment]
actually, could you provide a log with drachtio server at loglevel debug and sofia at loglevel 9?

**Comment by spointer** [comment]
I sent the log to your email.
Thank you.

**Comment by davehorton** [comment]
I actually see a different problem here.

The scenario looks like this:

UAC => P1 => P2 => UAS

i.e the UAC sends the INVITE through 2 proxies to the ultimate destination.
You have provided the log for proxy P2, which is the drachtio server.

When the first INVITE arrives to P2, it has a Record-Route header of P1, and two Via headers, as expected:
```
2019-10-22 00:45:34.800966 recv 888 bytes from udp/[111.111.11.247]:5060 at 00:45:34.800626:
INVITE sip:07080942391@121.160.50.106:5060;user=phone SIP/2.0
Via: SIP/2.0/UDP 111.111.11.247:5060;branch=z9hG4bK1d3e92ab45dae512e-54151-0_344401
Via: SIP/2.0/UDP 111.111.11.180:5060;branch=z9hG4bK1d3e92ab45dae512e-54151-0
Record-Route: <sip:111.111.11.247;lr>
```
P2, the drachtio server, adds its own RR and Via and sends it on to the destination you specified in your app:
```
2019-10-22 00:45:34.825428 send 1038 bytes to udp/[121.160.50.106]:3421 at 00:45:34.825086:
INVITE sip:07080942391@121.160.50.106:5060;user=phone SIP/2.0
Via: SIP/2.0/UDP 121.160.50.106:3417;rport;branch=z9hG4bK93366cb3-f5e9-42c1-a691-7b67b48865eb
Via: SIP/2.0/UDP 111.111.11.247:5060;branch=z9hG4bK1d3e92ab45dae512e-54151-0_344401;rport=5060
Via: SIP/2.0/UDP 111.111.11.180:5060;branch=z9hG4bK1d3e92ab45dae512e-54151-0
Record-Route: <sip:121.160.50.106:3417;lr>
Record-Route: <sip:111.111.11.247;lr>
```
Next, the 200 OK comes back to P2, which forwards it back upstream:
```
2019-10-22 00:45:36.387583 send 853 bytes to udp/[111.111.11.247]:5060 at 00:45:36.386995:
SIP/2.0 200 OK
Via: SIP/2.0/UDP 111.111.11.247:5060;branch=z9hG4bK1d3e92ab45dae512e-54151-0_344401;rport=5060
Via: SIP/2.0/UDP 111.111.11.180:5060;branch=z9hG4bK1d3e92ab45dae512e-54151-0
Record-Route: <sip:121.160.50.106:3417;lr>
Record-Route: <sip:111.111.11.247;lr>
```
Now comes the problem.  P1 forwards the ACK, but the Route header is wrong.  

P1 should construct the Route header based on the RR of the 200 OK that it received; in other words, the Route header should be `<sip:121.160.50.106:3417;lr>`.

But look at the ACK that P2 receives:
```
2019-10-22 00:45:36.496608 recv 514 bytes from udp/[111.111.11.247]:5060 at 00:45:36.496427:
ACK sip:07080942391@121.160.50.106:5060;user=phone SIP/2.0
Via: SIP/2.0/UDP 111.111.11.247:5060;branch=z9hG4bK1d3e92ab45dae512e-54151-1_344401
Via: SIP/2.0/UDP 111.111.11.180:5060;branch=z9hG4bK1d3e92ab45dae512e-54151-1
Route: <sip:121.160.50.106;lr>
```
It has dropped the port (`:3417`) entirely from the target.

Thus, P1 has misconstructed the target set.

P2 obeys the Route header and tries forwarding to that target, since it recognizes it is not a "local" address:

```
2019-10-22 00:45:36.504006 send 579 bytes to udp/[121.160.50.106]:5060 at 00:45:36.503700:
ACK sip:07080942391@121.160.50.106:5060;user=phone SIP/2.0
```
So, in summary, I think you have a problem with the proxy at P1 that caused the issue here

**Comment by spointer** [comment]
OK, i will look into the wrong route header.
By the way, if the route header is correct(ie, 3417 not omitted), i wonder how can P2 proxy to UAS correctly?(The port of UAS is 3421)

**Comment by davehorton** [comment]
P2 will proxy the ACK to UAS because the 200 OK Contact header had that IP and port:
```
2019-10-22 00:45:36.384054 recv 974 bytes from udp/[121.160.50.106]:3421 at 00:45:36.383834:
SIP/2.0 200 OK
Via: SIP/2.0/UDP 121.160.50.106:3417;rport=42258;branch=z9hG4bK93366cb3-f5e9-42c1-a691-7b67b48865eb;received=172.26.0.1
Via: SIP/2.0/UDP 111.111.11.247:5060;branch=z9hG4bK1d3e92ab45dae512e-54151-0_344401;rport=5060
Via: SIP/2.0/UDP 111.111.11.180:5060;branch=z9hG4bK1d3e92ab45dae512e-54151-0
Record-Route: <sip:121.160.50.106:3417;lr>
Record-Route: <sip:111.111.11.247;lr>
Contact: <sip:121.160.50.106:3421>
```
Thus, if the Route header on the ACK were correct when received by P2, it would be forwarding the ACK to 21.160.50.106:3421.

**Comment by spointer** [comment]
Thank you for your explanation but i wonder if the contact header from UAS 200 OK response is correct, wouldn't it be possible to send Ack regardless of the fact that the route header is wrong?

**Comment by davehorton** [comment]
No, RFC 3261 is very specific about how a proxy is to route requests within a dialog when Route header is present. Any other behavior would be not following the standard

**Comment by spointer** [comment]
Thank you very much and i close this issue.
